include/nbl/asset/filters/kernels/WeightFunctions.h

```diff
@@ -0,0 +1,308 @@
+#ifndef _NBL_ASSET_MATH_FUNCTIONS_H_INCLUDED_
+#define _NBL_ASSET_MATH_FUNCTIONS_H_INCLUDED_
+
+#include <limits>
+#include <cstdint>
+
+namespace nbl::asset
+{
+
+struct SDiracFunction
+{
+	// We can use std::nextafter after C++23
+	constexpr static inline float min_support = -1e-6f;
+	constexpr static inline float max_support = +1e-6f;
+	constexpr static inline uint32_t k_smoothness = 0;
+
+	template<int32_t derivative=0>
+	static inline double weight(float x)
+	{
+		if (x != 0.f)
+			return 0.0;
+
+		if constexpr (derivative == 0)
+		{
+			return std::numeric_limits<double>::infinity();
+		}
+		else
+		{
+			static_assert(false);
+			return core::nan<double>();
+		}
+	}
+};
+
+// Standard Box function, symmetric, value in the support is 1, integral is 1, so support must be [-1/2,1/2)
+struct SBoxFunction
+{
+	constexpr static inline float min_support = -0.5f;
+	constexpr static inline float max_support = +0.5f;
+	constexpr static inline uint32_t k_smoothness = 0;
+
+	template <int32_t derivative = 0>
+	static inline double weight(float x)
+	{
+		if (x >= min_support && x < max_support)
+		{
+			if constexpr (derivative == 0)
+			{
+				return 1.0;
+			}
+			else
+			{
+				static_assert(false);
+				return core::nan<double>();
+			}
+		}
+		return 0.0;
+	}
+};
+
+// Standard Triangle function, symmetric, peak in the support is 1 and at origin, integral is 1, so support must be [-1,1)
+struct STriangleFunction
+{
+	constexpr static inline float min_support = -1.f;
+	constexpr static inline float max_support = +1.f;
+	constexpr static inline uint32_t k_smoothness = 0;
+
+	template <int32_t derivative = 0>
+	static inline double weight(float x)
+	{
+		if (x >= min_support && x < max_support)
+		{
+			if constexpr (derivative > 0)
+			{
+				// Derivative at 0 not defined.
+				static_assert(false);
+				return core::nan<double>();
+			}
+			else
+			{
+				return 1.0 - core::abs(x);
+			}
+		}
+		return 0.0;
+	}
+};
+
+// Truncated Gaussian function, with stddev = 1.0, if you want a different stddev then you need to scale it.
+struct SGaussianFunction
+{
+	constexpr static inline float min_support = -3.f;
+	constexpr static inline float max_support = +3.f;
+	constexpr static inline uint32_t k_smoothness = std::numeric_limits<uint32_t>::max();
+
+	template <int32_t derivative = 0>
+	static inline double weight(float x)
+	{
+		if (x >= min_support && x < max_support)
+		{
+			if constexpr (derivative == 0)
+			{
+				const double normalizationFactor = core::inversesqrt(2.0 * core::PI<double>()) / std::erff(core::sqrt<double>(2.0) * double(min_support));
+				return normalizationFactor * exp2(-0.72134752 * x * x);
+			}
+			else if constexpr (derivative == 1)
+			{
+				return -x * SGaussianFunction::weight(x);
+			}
+			else if constexpr (derivative == 2)
+			{
+				return x * (x + 1.0) * SGaussianFunction::weight(x);
```

Comment:
I think product rule dictates this should be `(x*x-1)*Gauss` so you're wrong here

it would be safest for you to define higher order derivatives "compile time recursively"
```cpp
template<uint32_t order>
class polynomial_t final : public std::array<int64_t,order+1>
{
        using base_t = std::array<int64_t,order+1>;

    public:
        constexpr polynomial_t() : base_t() {}
        constexpr polynomial_t(base_t _list) : base_t(_list) {}

        template<typename T>
        constexpr T evaluate(const T x) const
        {
            T result = T(base_t::back());
            for (int64_t i=order-1; i>=0ll; i--)
            {
                result *= x;
                result += base_t::operator[](i);
            }
            return result;
        }

        constexpr auto differentiate() const
        {
            polynomial_t<order-1> f_prime;
            for (uint32_t power=1; power<=order; power++)
                f_prime[power-1] = base_t::operator[](power)*power;
            return f_prime;
        }
};
```

then in your gaussian's weight you could have
```cpp
private:
    template<int32_t derivative>
    constexpr static polynomial_t<derivative> differentialPolynomialFactor()
    {
        static_assert(derivative>0,"DIFFERENTIATION ONLY!");
        if constexpr (derivative!=1)
        {
            constexpr auto f = differentialPolynomialFactor<derivative-1>();
            constexpr auto f_prime = f.differentiate();
            polynomial_t<derivative> retval;
            for (uint32_t power=0; power<f_prime.size(); power++)
                retval[power] = f_prime[power];
            for (uint32_t power=0; power<f.size(); power++)
                retval[power+1] -= f[power];
            return retval;
        }
        else
            return std::array<int64_t,2>{0,-1};
    }
```
and implement your `derivative>0` as
```cpp
else if constexpr (derivative>0)
   return differentialPolynomialFactor<derivative>().evaluate(x)*SGaussianFunction::weight(x);
```

Reply:
https://godbolt.org/z/YchrEj1Tj

Reply:
solved it myself